### PR TITLE
monitoring(frontend): use ratios instead of hard thresholds

### LINKED ISSUE #12158: frontend: hard errors alerts as ratio/percentage instead of absolute-value
Writeup by @slimsag in https://github.com/sourcegraph/sourcegraph/issues/12011 re: `frontend: hard_error_search_responses`

---

It's not clear to me if this one is a fluke or not, because [sourcegraph.com has some quite bad search perf issues we haven't been following up on](https://github.com/sourcegraph/sourcegraph/issues/9359) - but a more direct query shows me the percentage of requests resulting in a hard error here is low so I presume we do not care and would like to relax the alert to account for a percentage instead of an absolute number of failing requests.

[this query](https://sourcegraph.com/-/debug/grafana/explore?orgId=1&left=%5B%22now-7d%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22sum%20by%20(source,request_name)(increase(src_graphql_search_response%7Bstatus%3D~%5C%22error%5C%22%7D%5B1d%5D))%22,%22legendFormat%22:%22hard%20error%22,%22datasource%22:%22Prometheus%22%7D,%7B%22mode%22:%22Metrics%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D&right=%5B%22now-7d%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22sum%20by%20(source,request_name)(increase(src_graphql_search_response%7Bstatus!~%5C%22error%5C%22%7D%5B1d%5D))%22,%22legendFormat%22:%22hard%20error%22,%22datasource%22:%22Prometheus%22%7D,%7B%22mode%22:%22Metrics%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D)

shows that the cause is about 300 browser code intel requests per day (find-references, jump-to-def, etc.) end in a hard error, where the user gets an error back instead of a result. If I had to guess, these are something like a timeout edge case we aren't recording properly (e.g. the user navigates away from the page while it is in flight), or correlated with deployments of services.

While that is worrying, the same query above shows that only accounts for only around 0.1% of requests soI think the TODO here is two-fold:

1. File an issue to track turning this into a ratio/percentage instead of absolute-value alert.
2. Silence the alert for now on Sourcegraph.com

### PR DESCRIPTION
**Pull request by bobheadxi**
While the original alert described in #12158 is not as noisy anymore, in general some of the noisiest alerts in `#alerts-cloud` are those alerts with hard thresholds, ie "_Y_+ errors in _X_ minutes" - on larger instances like Sourcegraph Cloud, this could mean we fire alerts on issues that only affect a very small number of users.

This change tries to convert all of the frontend service's hard threshold alerts to ratio-based alerts. I opted to just change them all for consistency.

To try and account for smaller instances where a few errors in a few requests can trigger an alert, I've added a `for` parameter to each alert, so that requests should consistently error out before causing an alert to be raised.

Closes #12158 

Some comparisons - left is previous query, right is new ratio query. Each link also has a second metric query on the old panel to compare with total requests

* [`hard_error_search_responses`](https://sourcegraph.com/-/debug/grafana/explore?orgId=1&left=%5B%22now-7d%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22sum%20by%20(status)(increase(src_graphql_search_response%7Bstatus%3D~%5C%22error%5C%22,source%3D%5C%22browser%5C%22,name!%3D%5C%22CodeIntelSearch%5C%22%7D%5B5m%5D))%22,%22legendFormat%22:%22hard%20error%22,%22datasource%22:%22Prometheus%22%7D,%7B%22mode%22:%22Metrics%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D&right=%5B%22now-7d%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22sum%20by%20(status)(increase(src_graphql_search_response%7Bstatus%3D~%5C%22error%5C%22,source%3D%5C%22browser%5C%22,name!%3D%5C%22CodeIntelSearch%5C%22%7D%5B5m%5D))%20%2F%20ignoring(status)%20group_left%20sum(increase(src_graphql_search_response%7Bsource%3D%5C%22browser%5C%22,name!%3D%5C%22CodeIntelSearch%5C%22%7D%5B5m%5D))%20*%20100%22,%22legendFormat%22:%22hard%20error%22,%22datasource%22:%22Prometheus%22%7D,%7B%22mode%22:%22Metrics%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D)

image

* [`precise_code_intel_api_errors`](https://sourcegraph.com/-/debug/grafana/explore?orgId=1&left=%5B%22now-2d%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22sum(increase(src_code_intel_api_errors_total%5B5m%5D))%22%7D,%7B%22expr%22:%22sum(increase(src_code_intel_api_total%5B5m%5D))%22,%22hide%22:true%7D,%7B%22mode%22:%22Metrics%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D&right=%5B%22now-2d%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22sum(increase(src_code_intel_api_errors_total%5B5m%5D))%20%2F%20sum(increase(src_code_intel_api_total%5B5m%5D))%20*%20100%22%7D,%7B%22mode%22:%22Metrics%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D) - this one is [particularly noisy](https://sourcegraph.com/-/debug/grafana/explore?orgId=1&left=%5B%22now-2d%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22alert_count%7Bname%3D%5C%22precise_code_intel_api_errors%5C%22%7D%22%7D,%7B%22mode%22:%22Metrics%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D)

image

If this change seems useful, I can apply it to our other services as well

### CODE REVIEW

File: doc/admin/observability/alert_solutions.md [hunk 1]
```diff
@@ -331,9 +331,9 @@ To learn more about Sourcegraph's alerting, see [our alerting documentation](htt
 
 **Descriptions:**
 
-- _frontend: 5+ hard error search API responses every 5m_
+- _frontend: 5%+ hard error search API responses every 5m_
 
-- _frontend: 20+ hard error search API responses every 5m_
+- _frontend: 20%+ hard error search API responses every 5m_
```
Comment (tsenart) [inline doc/admin/observability/alert_solutions.md:336]: Isn't this a bit too high? Shouldn't we be alerted before we reach 20% hard error rate? Similar question for all the other conversions to rates. Do they need a bit more thought?

Comment (bobheadxi) [reply doc/admin/observability/alert_solutions.md:336]: What do you think would be more suitable? I'm not sure what kind of behaviour the original thresholds were supposed to fire on - should these be something closer to 5% warning, 10% critical?

Comment (tsenart) [reply doc/admin/observability/alert_solutions.md:336]: I'd say 2% warning and 5% critical, but that's a very uninformed guess. What alerts would that result in today?

Comment (bobheadxi) [reply doc/admin/observability/alert_solutions.md]: Just eyeballing the new thresholds in f0e4db99c83836a3a98e63aa54d85a6447b8be0f, the only one I see that is likely to trigger is `internal_unindexed_search_error_responses`:

image

none of the spikes persist for 15minutes, however - everything else seems clear

Comment (bobheadxi) [reply doc/admin/observability/alert_solutions.md]: Could 15 minutes be too long a requirement?

**Pull request merged**